### PR TITLE
feat(jans-cedarling): implement `KeyService` for `JwtService` for key management

### LINKED ISSUE #9814: feat(jans-cedarling): implement a `KeyService` for `JwtService`
**Is your feature request related to a problem? Please describe.**

The `JwtService` needs decoding keys to validate JWTs. A key `KeyService` should be implemented to manage said keys.

**Describe the solution you'd like**

A `KeyService` would be implemented which will be responsible for the following:

- Fetch keys from the `jwks_uri`
- provide keys for the `JwtService` to use

**Describe alternatives you've considered**

**Additional context**

This will be a dependency for `JwtService` to work with validation turned on.

### PR DESCRIPTION
**Pull request by rmarinn**
***IMPORTANT:*** merge #9910 before merging this.

### Prepare

- [x] Read [PR guidelines](https://github.com/JanssenProject/jans/blob/main/docs/CONTRIBUTING.md#prs)
- [x] Read [license information](https://github.com/JanssenProject/jans/blob/main/LICENSE)

-------------------

### Description

This PR introduces a new `KeyService` for the `JwtService`, which is responsible for the following core functionalities:
- Fetching Decoding Keys: The `KeyService` retrieves decoding keys from a JSON Web Key Set (JWKS) provided by Identity Providers (IDPs). This allows the JwtService to validate JWTs effectively.
- Storing Decoding Keys: The `KeyService` exposes a `get_key` function that checks for locally stored keys. If no keys are found, the `KeyService` attempts to fetch the keys from the jwks_uri again before returning an error.

#### Target issue
 
The implementation addresses the need for a dedicated service to manage the decoding keys necessary for JWT validation, as highlighted in issue #9814.

closes #9814 

#### Implementation Details

The `KeyService` is designed to:
- Connect to specified JWKS endpoints to retrieve and cache decoding keys.
- Handle errors related to key retrieval and provide clear feedback on issues encountered.
- Support efficient key management, ensuring that keys are kept up-to-date and available for validation processes.

-------------------
### Test and Document the changes
- [x] Static code analysis has been run locally and issues have been fixed
- [x] Relevant unit and integration tests have been added/updated
- [x] Relevant documentation has been updated if any (i.e. user guides, installation and configuration guides, technical design docs etc)

Please check the below before submitting your PR. The PR will not be merged if there are no commits that start with `docs:` to indicate documentation changes or if the below checklist is not selected.
- [x] **I confirm that there is no impact on the docs due to the code changes in this PR.**

### CODE REVIEW

**Comment by rmarinn** [comment]
this should be functionally OK but breaks the following tests which are unreleted to the jwt module:
- `init::cedar_schema::tests::test_read_ok`
- `init::cedar_schema::tests::test_read_ok`

the tests are breaking because of adding `trusted_idps` to the schema:
```rust
// src/models/policy_store.rs

#[derive(Debug, Clone, serde::Deserialize)]
pub struct PolicyStore {
    #[serde(deserialize_with = "parse_cedar_schema")]
    #[allow(dead_code)]
    pub schema: CedarSchema,
    #[serde(deserialize_with = "parse_policy_set")]
    #[allow(dead_code)]
    pub policies: cedar_policy::PolicySet,
    pub trusted_idps: Vec<TrustedIdp>, // adding this line breaks the tests
}
```

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: It's better to use `eprintln!`

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: You can directly use the client instead of `get`, as you mention in your TODO 
```
      let client = Client::new();
      let mut idp_configs = HashMap::new();

      // fetch IDP configs
      for endpoint in openid_conf_endpoints {
          let conf_src: OpenIdConfigSource = client.get(endpoint)
              .send()
              .map_err(Error::HttpError)?
              .error_for_status()
              .map_err(Error::HttpError)?
              .json()
              .map_err(Error::RequestDeserializationError)?;
          let (issuer, conf) = OpenIdConfig::from_source(conf_src);
          idp_configs.insert(issuer, conf);
      }
```

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: and here 
```
client.get(&*conf.jwks_uri)
.send()
```

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: `use reqwest::blocking::Client;`

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs:86]: Is it really useful to clone here ?

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs:93]: Same here, is it really useful to clone ?

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs:20]: maybe it can be useful to add 
 `client: Client,`
in the struct to use it in `new` and `update_jwks`.

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: and return `Ok(Self { idp_configs, client })`

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: and here you can use 
`self.client.get(&*conf.jwks_uri).send()`

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: maybe you can manage better the error here

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (abaghinyan) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: maybe you can manage better the error here

**Review by abaghinyan — CHANGES_REQUESTED** [review]
Some suggestion of modification for `key_service`

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs:86]: I am cloning an `Arc<DecodingKey>` here, not the whole decoding key. I am doing this so that the call to `get_key` will not require the reference to `KeyService` to be mutable, e.g. `get_key(&mut self, ...)`.

Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs:86]: note that calls to `get_key` updates the local keystore if a key could not be found.

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: resolved by: [7976b58](https://github.com/JanssenProject/jans/pull/9855/commits/7976b5840cfe5ba7851800dc95325b006f1de8ef)

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: resolved by: [7976b58](https://github.com/JanssenProject/jans/pull/9855/commits/7976b5840cfe5ba7851800dc95325b006f1de8ef)

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs:20]: resolved by: [7976b58](https://github.com/JanssenProject/jans/pull/9855/commits/7976b5840cfe5ba7851800dc95325b006f1de8ef)

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: resolved by: [7976b58](https://github.com/JanssenProject/jans/pull/9855/commits/7976b5840cfe5ba7851800dc95325b006f1de8ef)

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: resolved by: [7976b58](https://github.com/JanssenProject/jans/pull/9855/commits/7976b5840cfe5ba7851800dc95325b006f1de8ef)

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: resolved by: [7976b58](https://github.com/JanssenProject/jans/pull/9855/commits/7976b5840cfe5ba7851800dc95325b006f1de8ef)

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: resolved by: [7976b58](https://github.com/JanssenProject/jans/pull/9855/commits/7976b5840cfe5ba7851800dc95325b006f1de8ef)

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs:86]: Here's a list of the reasons why I decided to it this way:
- we're probably going to switch to using Async in the future so this keeps it thread-safe
- no need to specify lifetimes which is a pain in Async/multi-threaded applications. (i'm not sure yet if we're going to implement multi-threading but it has been mentioned)
- `update_jwks` potentially modifies the internal key store (by fetching new keys), returning an Arc avoids issues where a reference to a key might become invalid after the key store is updated. Arc guarantees that the key remains valid as long as it's still being used, regardless of updates to the key store.

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: added a custom error. see: [460bc0b](https://github.com/JanssenProject/jans/pull/9855/commits/460bc0bf241f11a896fa6c8482bd72e0f1e353dd)

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs]: added a custom error. see: [460bc0b](https://github.com/JanssenProject/jans/pull/9855/commits/460bc0bf241f11a896fa6c8482bd72e0f1e353dd)

**Review by nynymike — DISMISSED** [review]

**Review by olehbozhok — DISMISSED** [review]

**Review by duttarnab — DISMISSED** [review]

**Review by abaghinyan — DISMISSED** [review]

**Comment by rmarinn** [comment]
wait for #9910 to merge before merging this.

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (djellemah) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs:24]: This silences a couple of unused variable warnings - `iss` in `for (iss, conf`, and `config` in `for (iss, config` in `fn get_key`.

Also it hides an unused `Result` from `self.update_jwks(iss)`

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (djellemah) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs:88]: Seems to me this should end up in a log somewhere, rather than on stderr.

Perhaps that change should be deferred until after this PR is accepted.

File: jans-cedarling/cedarling/src/jwt/decoding_strategy.rs
Comment (djellemah) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy.rs:172]: Could be a little nicer at the call in `parse_jwt_algorithms` if you used this signature
```
string_to_alg<S : AsRef<str>>(algorithm: S)
```

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service/error.rs
Comment (djellemah) [inline jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service/error.rs:12]: What is the external name for a key id?

**Review by djellemah — APPROVED** [review]
A couple of general comments:
- `jwt/decoding_strategy/traits.rs` seems to be unused.
- `dead_code` is hiding some struct members that are unused. I'm thinking of `TrustedIssuer` `TokenMetadata` and `TokenKind`.

Those might be intentional, and I don't see my other comments as blockers, so I'm approving the PR anyway.

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs:88]: i added a "TODO" string somewhere in PR #9939 about that... sorry it's such a mess right now, i'm planning to do a refactor which fixes a lot of the issues once the other PRs merges

File: jans-cedarling/cedarling/src/jwt/decoding_strategy.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy.rs:172]: okay, i'll keep that in mind for the refactor

File: jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service/error.rs
Comment (rmarinn) [reply jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service/error.rs:12]: im not sure what you mean by external name but JSON Web Keys could optionally have a `kid` field which refers to the *key id* of the key.

**Pull request merged**
